### ISSUE TITLE
Problem replicating more than one set of snapshots FreeBSD 11.1

### ISSUE THREAD

**Issue by ssnor** [Jehops/zap#7]
Hello!

I'm having trouble implementing this script on my home server.
Following along with the blogpost recommended here I reach a snag when I try to replicate more than one set of snapshots.
When I create snapshots with e.g. 'zap snap 1d' and then replicate it with 'zap rep -v' it works just fine. But if i create another snapshot and try to replicate again zfs spits out this error: "cannot receive new filesystem stream: destination has snapshots ..."

I have attached a file [here](https://github.com/Jehops/zap/files/1272801/zap_error.txt) with all the outputs and some information about my pools such as permissions for user zap and the properties zap:snap and snap:name

I have tried this with both version 0.7.1 (from ports) and 0.7.2 (I just copied the script from here) but nothing works.

If I manually try something like: `zap@serv1BSD:/usr/home/dano % zfs send -p -i zroot/var/audit@ZAP_serv1BSD_2017-09-02T13:50:10p0200--1w zroot/var/audit@ZAP_serv1BSD_2017-09-02T14:00:44p0200--1d | ssh zap@localhost "zfs recv -Fu -v -d zstore/zroot_bak"`

It works, but the output shows some errors about permissions:
`receiving incremental stream of zroot/var/audit@ZAP_serv1BSD_2017-09-02T14:00:44p0200--1d into zstore/zroot_bak/var/audit@ZAP_serv1BSD_2017-09-02T14:00:44p0200--1d
cannot receive setuid property on zstore/zroot_bak/var/audit: permission denied
cannot receive exec property on zstore/zroot_bak/var/audit: permission denied
received 312B stream in 1 seconds (312B/sec)`

So now I'm at my wits end and don't know what to do, but I'm sure I have managed to misconfigure something.

Thank you for `zap`!

**Comment by Jehops** [comment]
Hi.

I just pushed a version of `zap` to the `ssnor` branch, which will print some debugging information like below.  Could you try it and share the output?  Could you also list the snapshots for both pools between each step?  I have not been able to reproduce the problem you are experiencing, but the `Ambiguous output redirect.` that only you get looks suspicious.

```
$ ./zap rep -v zap@localhost:zstore/zroot_bak zroot/tmp
Sun Sep  3 10:49:13 ADT 2017
Replicating...
DEBUG: ****************** rep argument: zap@localhost:zstore/zroot_bak *********************
DEBUG: ****************** sshto: zap@localhost *******************
DEBUG: ****************** rloc: zstore/zroot_bak *******************
DEBUG: ****************** l_ts: 1504446551 *******************
DEBUG: ****************** fs: tmp *******************
DEBUG: ****************** rsnap: @ZAP_phe_2017-09-03T10:44:28-0300--1d *******************
zfs send -I zroot/tmp@ZAP_phe_2017-09-03T10:44:28-0300--1d zroot/tmp@ZAP_phe_2017-09-03T10:49:11-0300--1d | ssh zap@localhost "zfs recv -du  -v zstore/zroot_bak"
receiving incremental stream of zroot/tmp@ZAP_phe_2017-09-03T10:49:11-0300--1d into zstore/zroot_bak/tmp@ZAP_phe_2017-09-03T10:49:11-0300--1d
received 108KB stream in 1 seconds (108KB/sec)
zfs bookmark zroot/tmp@ZAP_phe_2017-09-03T10:49:11-0300--1d zroot/tmp#ZAP_phe_2017-09-03T10:49:11-0300--1d
Created bookmark for zroot/tmp@ZAP_phe_2017-09-03T10:49:11-0300--1d.
```

**Comment by ssnor** [comment]
Here is the output from the debug branch: [zap_debug_output.txt](https://github.com/Jehops/zap/files/1273206/zap_debug_output.txt)
Something else that's weird is that if I try to do a replication of an individual snapshot like in your example above this is my [output.](https://github.com/Jehops/zap/files/1273216/zap_single_snapshot_replication.txt)

As you can see it says that it is replicating but nothing happens.
I hope that helps, just say the word if you need anything else.

**Comment by Jehops** [comment]
Could you show the output of `ssh zap@localhost "zfs list -rd1 -tsnap -o name -s creation zstore/zroot_bak/ROOT 2>/dev/null"`?

Note: 1. I initially forgot the closing `"`.  2. Please do this after `zap snap...`.

**Comment by ssnor** [comment]
Ok, I deleted all the snapshots and did this:
`./zap snap -v 1d`
`./zap rep -v`
`./zap snap -v 1w`
`./zap rep -v`

Then I tried your line above and got this:
`zap@serv1BSD:/usr/home/dano/bin/zap-ssnor % ssh zap@localhost "zfs list -rd1 -tsnap -o name -s creation zstore/zroot_bak/ROOT 2>/dev/null"
cannot open '2': dataset does not exist`
Removed the `2>/dev/null` and got this:
`zap@serv1BSD:/usr/home/dano/bin/zap-ssnor % ssh zap@localhost "zfs list -rd1 -tsnap -o name -s creation zstore/zroot_bak/ROOT"
NAME
zstore/zroot_bak/ROOT@ZAP_serv1BSD_2017-09-04T04:02:46p0200--1d`

**Comment by Jehops** [comment]
Is the zap account still using `/bin/sh` as its shell?  If not, do you see the same thing with this?
```
% sh
% ssh zap@localhost "zfs list -rd1 -tsnap -o name -s creation zstore/zroot_bak/ROOT 2>/dev/null"
```

**Comment by ssnor** [comment]
I figured out the problem!

As I suspected it was my own misconfiguration at fault.
I had changed the default shell for user zap to `tcsh` because I was trying to use aliases when setting up `ssh` to localhost and `sh` won't load ~/.shrc in a non-interactive shell by default while `tcsh` does. 

I ended up using ~/.ssh/config to solve that the proper way, but I didn't bother to change the default shell back because the script calls for `/bin/sh` anyway so I thought it wouldn't make a difference.
I guess the problem stems from some quirk in the `tcsh` syntax that broke something.
When I was using `tcsh` as default shell `zap` uses `zfs send -p...` when replicating to a pool with existing snapshots but with `sh` it uses `zfs send -I...`.
I would recommend you to add the required default shell for user zap to the manpage.

Thank you so much for your help, now I'm off to edit my crontab :)
